### Data Preparation Project. 
Scenario:   
You have been retained by a haulage company to analyse a dataset based on data collected from heavy Scania trucks in everyday usage. The system in focus is the Air Pressure system (APS) which generates pressurised air that are utilized in various functions in a truck, such as braking and gear changes. The dataset’s  positive class consists of component failures for a specific component of the APS system.  The negative class consists of trucks with failures for components not related to the APS. The data consists  of a subset of all available data, selected by experts. This analysis will help determine the investment strategy for the company in the upcoming year.   



In [18]:
%matplotlib inline

In [19]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns; 
import warnings 
warnings.filterwarnings('ignore') 
sns.set() 

In [20]:
aps_failure = pd.read_csv("aps_failure_set.csv")
aps_failure.head()

class  aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0   neg   76698     na  2130706438    280      0      0      0      0      0   
1   neg   33058     na           0     na      0      0      0      0      0   
2   neg   41040     na         228    100      0      0      0      0      0   
3   neg      12      0          70     66      0     10      0      0      0   
4   neg   60874     na        1368    458      0      0      0      0      0   

   ...   ee_002  ee_003  ee_004  ee_005  ee_006  ee_007  ee_008 ee_009 ef_000  \
0  ...  1240520  493384  721044  469792  339156  157956   73224      0      0   
1  ...   421400  178064  293306  245416  133654   81140   97576   1500      0   
2  ...   277378  159812  423992  409564  320746  158022   95128    514      0   
3  ...      240      46      58      44      10       0       0      0      4   
4  ...   622012  229790  405298  347188  286954  311560  433954   1218      0   

  eg_000  
0      0  
1      0  
2      0  
3     32  
4      0  

[5 rows x 171 columns]

In [21]:
aps_failure.isnull().sum()

class     0
aa_000    0
ab_000    0
ac_000    0
ad_000    0
         ..
ee_007    0
ee_008    0
ee_009    0
ef_000    0
eg_000    0
Length: 171, dtype: int64

### STEP 1 
Using .head to see the first five observations we realized that there are missing values. 
However, these missing values are not NaN or empty, it is fill with na, we need to change it, instance to be able to continue our process. 

In [22]:
aps_failure.shape

(60000, 171)

To be able to see how much data we are working with, I used .shape and as we can see it there are 60.000 observations and 171 features. 

In [23]:
aps_failure.describe()

aa_000
count  6.000000e+04
mean   5.933650e+04
std    1.454301e+05
min    0.000000e+00
25%    8.340000e+02
50%    3.077600e+04
75%    4.866800e+04
max    2.746564e+06

In [24]:
aps_failure.dtypes

class     object
aa_000     int64
ab_000    object
ac_000    object
ad_000    object
           ...  
ee_007    object
ee_008    object
ee_009    object
ef_000    object
eg_000    object
Length: 171, dtype: object

In [25]:
aps_failure.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Columns: 171 entries, class to eg_000
dtypes: int64(1), object(170)
memory usage: 78.3+ MB


Upon we can see that the Data types are int64 and Object.
However, I need numeric. 


### ADD OBSERVATION HERE ABOUT CLEAN DATA. 

In [26]:
aps_failure.head()

class  aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0   neg   76698     na  2130706438    280      0      0      0      0      0   
1   neg   33058     na           0     na      0      0      0      0      0   
2   neg   41040     na         228    100      0      0      0      0      0   
3   neg      12      0          70     66      0     10      0      0      0   
4   neg   60874     na        1368    458      0      0      0      0      0   

   ...   ee_002  ee_003  ee_004  ee_005  ee_006  ee_007  ee_008 ee_009 ef_000  \
0  ...  1240520  493384  721044  469792  339156  157956   73224      0      0   
1  ...   421400  178064  293306  245416  133654   81140   97576   1500      0   
2  ...   277378  159812  423992  409564  320746  158022   95128    514      0   
3  ...      240      46      58      44      10       0       0      0      4   
4  ...   622012  229790  405298  347188  286954  311560  433954   1218      0   

  eg_000  
0      0  
1      0  
2      0  
3     32  
4      0  

[5 rows x 171 columns]

### Fist of all to clean the data we are using the code upon to calc the mean and the line above we are replacing 'na' to Mean. 

I am going to replace any 'na' value to NaN and then I'll drop all the columns with missing values. 

In [35]:
aps_failure_new = aps_failure.replace('na',np.nan)


In [36]:
aps_failure_new

class  aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001  \
0       neg   76698    NaN  2130706438    280      0      0      0      0   
1       neg   33058    NaN           0    NaN      0      0      0      0   
2       neg   41040    NaN         228    100      0      0      0      0   
3       neg      12      0          70     66      0     10      0      0   
4       neg   60874    NaN        1368    458      0      0      0      0   
...     ...     ...    ...         ...    ...    ...    ...    ...    ...   
59995   neg  153002    NaN         664    186      0      0      0      0   
59996   neg    2286    NaN  2130706538    224      0      0      0      0   
59997   neg     112      0  2130706432     18      0      0      0      0   
59998   neg   80292    NaN  2130706432    494      0      0      0      0   
59999   neg   40222    NaN         698    628      0      0      0      0   

      ag_002  ...   ee_002  ee_003   ee_004   ee_005   ee_006  ee_007  ee_008  \
0          0  ...  1240520  493384   721044   469792   339156  157956   73224   
1          0  ...   421400  178064   293306   245416   133654   81140   97576   
2          0  ...   277378  159812   423992   409564   320746  158022   95128   
3          0  ...      240      46       58       44       10       0       0   
4          0  ...   622012  229790   405298   347188   286954  311560  433954   
...      ...  ...      ...     ...      ...      ...      ...     ...     ...   
59995      0  ...   998500  566884  1290398  1218244  1019768  717762  898642   
59996      0  ...    10578    6760    21126    68424      136       0       0   
59997      0  ...      792     386      452      144      146    2622       0   
59998      0  ...   699352  222654   347378   225724   194440  165070  802280   
59999      0  ...   440066  183200   344546   254068   225148  158304  170384   

       ee_009 ef_000 eg_000  
0           0      0      0  
1        1500      0      0  
2         514      0      0  
3           0      4     32  
4        1218      0      0  
...       ...    ...    ...  
59995   28588      0      0  
59996       0      0      0  
59997       0      0      0  
59998  388422      0      0  
59999     158      0      0  

[60000 rows x 171 columns]

I've changed all the columns with values 'na' to 'NaN', Now, I can continue the process to drop all the observations with 

In [37]:
aps_failure_new.dtypes

class     object
aa_000     int64
ab_000    object
ac_000    object
ad_000    object
           ...  
ee_007    object
ee_008    object
ee_009    object
ef_000    object
eg_000    object
Length: 171, dtype: object

In [42]:
aps_failure_new.isnull().sum()

class         0
aa_000        0
ab_000    46329
ac_000     3335
ad_000    14861
          ...  
ee_007      671
ee_008      671
ee_009      671
ef_000     2724
eg_000     2723
Length: 171, dtype: int64

In [44]:
new_aps = aps_failure_new.dropna(axis=0)

In [45]:
new_aps.isnull().sum()

class     0
aa_000    0
ab_000    0
ac_000    0
ad_000    0
         ..
ee_007    0
ee_008    0
ee_009    0
ef_000    0
eg_000    0
Length: 171, dtype: int64

In [46]:
new_aps.shape

(591, 171)

In [47]:
new_aps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 591 entries, 16 to 59950
Columns: 171 entries, class to eg_000
dtypes: int64(1), object(170)
memory usage: 794.2+ KB


In [48]:
new_aps.head()

class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000  ag_001  ag_002  \
16    neg   31300      0    784    740      0      0      0       0       0   
179   neg   97000      0    378    160      0      0      0       0       0   
225   neg  124656      2    278    170      0      0      0       0       0   
394   pos  281324      2   3762   2346      0      0   4808  215720  967572   
413   pos   43482      0   1534   1388      0      0      0       0   40024   

     ...   ee_002  ee_003  ee_004   ee_005   ee_006  ee_007 ee_008 ee_009  \
16   ...   798872  112724   51736     7054     6628   27600      2      2   
179  ...  1078982  313334  511330   552328   871528  871104   1980     42   
225  ...  1205696  866148  697610   700400  1900386  437532   3680      0   
394  ...   624606  269976  638838  1358354   819918  262804   2824      0   
413  ...   497196  121166  202272   232636   645690      50      0      0   

    ef_000 eg_000  
16       0      0  
179      0      0  
225      0      0  
394      0      0  
413      0      0  

[5 rows x 171 columns]

In [60]:
new_aps['class'].replace(['pos','neg'],[1,2],inplace=True)

In [61]:
new_aps.head()

class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000  ag_001  ag_002  \
16       2   31300      0    784    740      0      0      0       0       0   
179      2   97000      0    378    160      0      0      0       0       0   
225      2  124656      2    278    170      0      0      0       0       0   
394      1  281324      2   3762   2346      0      0   4808  215720  967572   
413      1   43482      0   1534   1388      0      0      0       0   40024   

     ...   ee_002  ee_003  ee_004   ee_005   ee_006  ee_007 ee_008 ee_009  \
16   ...   798872  112724   51736     7054     6628   27600      2      2   
179  ...  1078982  313334  511330   552328   871528  871104   1980     42   
225  ...  1205696  866148  697610   700400  1900386  437532   3680      0   
394  ...   624606  269976  638838  1358354   819918  262804   2824      0   
413  ...   497196  121166  202272   232636   645690      50      0      0   

    ef_000 eg_000  
16       0      0  
179      0      0  
225      0      0  
394      0      0  
413      0      0  

[5 rows x 171 columns]

### STEP 2
We've already clean the data the decision that I made was. 
1. I've dropped all the columns with NaN, before that I changed 'na' values to NaN to be able to continue the process of data cleaning. 
2. I changed the colum class which there were two strings "pos' and 'neg' I replace to 1 = pos and 2 = neg. 